src/pkcs15init/pkcs15-oberthur.c

```diff
@@ -625,20 +625,25 @@ cosm_generate_key(struct sc_profile *profile, struct sc_pkcs15_card *p15card,
 	LOG_TEST_RET(ctx, rv, "Cannot generate key: 'CREATE' authentication failed");
 
 	sc_file_free(tmpf);
+	tmpf = NULL;
 
 	rv = sc_select_file(p15card->card, &key_info->path, &prkf);
 	LOG_TEST_RET(ctx, rv, "Failed to generate key: cannot select private key file");
 
 	/* In the private key DF create the temporary public RSA file. */
 	rv = cosm_get_temporary_public_key_file(p15card->card, prkf, &tmpf);
-	if (rv != SC_SUCCESS)
+	if (rv != SC_SUCCESS) {
 		sc_file_free(prkf);
-	LOG_TEST_RET(ctx, rv, "Error while getting temporary public key file");
+		sc_file_free(tmpf);
```

Comment:
I think this `sc_file_free` is not needed here. When error occurs in `cosm_get_temporary_public_key_file`, the allocated file should be freed there and `tmpf` remains `NULL`.

Reply:
Right. Removed that line. Thanks!